### PR TITLE
[Libomptarget] Remove requires information from plugin

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
Currently this is only used for the zero-copy handling. However, this
can easily be moved into `libomptarget` so that we do not need to bother
setting the requires flags in the plugin. The advantage here is that we
no longer need to do this for every device redundently. Additionally,
these requires flags are specifically OpenMP related, so they should
live in `libomptarget`.

### CODE REVIEW

**Comment by carlobertolli** [comment]
I need to look more carefully at the patch, but automatic zero copy decision is taken in the amdgpu plugin if the program did not have 'requires unified_shared_memory'. If you remove that information from the plugin how can it know that automatic zero-copy needs to be turned on?
At the same time, runtime decision on zero-copy for amdgpu's is based on the machine being an APU, and that is a vendor specific information that cannot be moved to libomptarget.

**Comment by jhuber6** [comment]
> I need to look more carefully at the patch, but automatic zero copy decision is taken in the amdgpu plugin if the program did not have 'requires unified_shared_memory'. If you remove that information from the plugin how can it know that automatic zero-copy needs to be turned on? At the same time, runtime decision on zero-copy for amdgpu's is based on the machine being an APU, and that is a vendor specific information that cannot be moved to libomptarget.

This was the only use upstream. I see there are some more uses downstream that I wasn't aware of. Obviously the APU stuff is device dependent which is why it stayed where it is.

The use downstream seems to be some kind of checks for prepopulating some table? I don't fully understand that. It seems you and @ThorBl were involved in those changes, so maybe you know what the requirements are there. Either way, I don't think that we should use the same "requires" flags.

**Comment by carlobertolli** [comment]
Current upstream implementation of auto zero-copy does not support global variables. They need to be handled differently from unified_shared_memory. Thinking about it, we should have all the information in libomptarget as we surfaced it as a requires flag. That is, in libomptarget we know that we are in auto zero-copy and not usm, and that should be sufficient to decide on how to handle global variables. 

I am attending the OpenMP face-2-face this week and I didn't have time to work on the patch to support global variables, but will do next week.

What we have downstream is support to prepopulate GPU page tables from the CPU without using XNACK. I haven't yet figured out if we need to exclude the usm case (prepopulating could apply to usm as well, as a potential optimization), so I don't yet know if we need the info in the plugin.

I *think* we can go on with this and still be able to support global variables for automatic zero-copy.

**Comment by jhuber6** [comment]
> Current upstream implementation of auto zero-copy does not support global variables. They need to be handled differently from unified_shared_memory. Thinking about it, we should have all the information in libomptarget as we surfaced it as a requires flag. That is, in libomptarget we know that we are in auto zero-copy and not usm, and that should be sufficient to decide on how to handle global variables.
> 
> I am attending the OpenMP face-2-face this week and I didn't have time to work on the patch to support global variables, but will do next week.
> 
> What we have downstream is support to prepopulate GPU page tables from the CPU without using XNACK. I haven't yet figured out if we need to exclude the usm case (prepopulating could apply to usm as well, as a potential optimization), so I don't yet know if we need the info in the plugin.
> 
> I _think_ we can go on with this and still be able to support global variables for automatic zero-copy.

Alright, sounds good. I'll leave you to it. Hope you're having fun at the F2F.

**Review by carlobertolli — APPROVED** [review]
Did you test on an MI300A? If not, please do before merging.

**Comment by jhuber6** [comment]
> Did you test on an MI300A? If not, please do before merging.

I don't think I have access to any currently. But I would expect just looking at the code upstream it would be inconsequential because I simply moved the `if` condition out of the plugins and into `libomptarget`. Thanks for the review, I'll hold off on merging it until we have a better solution downstream to make @ronlieb's life easier.

**Pull request merged**
